target/ppc/translate/vmx-impl.c.inc

```diff
@@ -1601,6 +1601,102 @@ static bool trans_VEXTRACTQM(DisasContext *ctx, arg_VX_tb *a)
     return true;
 }
 
+static bool do_mtvsrm(DisasContext *ctx, arg_VX_tb *a, unsigned vece)
+{
+    const uint64_t elem_length = 8 << vece, highest_bit = 15 >> vece;
+    int i;
+    TCGv_i64 t0, t1;
+
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VECTOR(ctx);
+
+    t0 = tcg_const_i64(0);
+    t1 = tcg_temp_new_i64();
+
+    for (i = 1 << highest_bit; i > 1 << (highest_bit / 2); i >>= 1) {
+        tcg_gen_shli_i64(t0, t0, 1);
+        tcg_gen_ext_tl_i64(t1, cpu_gpr[a->vrb]);
+        tcg_gen_andi_i64(t1, t1, i);
+        tcg_gen_setcond_i64(TCG_COND_NE, t1, t1, tcg_constant_i64(0));
+        tcg_gen_or_i64(t0, t0, t1);
+        tcg_gen_shli_i64(t0, t0, elem_length - 1);
+    }
+
+    set_avr64(a->vrt, t0, true);
+
+    for (; i > 0; i >>= 1) {
+        tcg_gen_shli_i64(t0, t0, 1);
+        tcg_gen_ext_tl_i64(t1, cpu_gpr[a->vrb]);
+        tcg_gen_andi_i64(t1, t1, i);
+        tcg_gen_setcond_i64(TCG_COND_NE, t1, t1, tcg_constant_i64(0));
+        tcg_gen_or_i64(t0, t0, t1);
+        tcg_gen_shli_i64(t0, t0, elem_length - 1);
+    }
+
+    set_avr64(a->vrt, t0, false);
+
+    tcg_gen_gvec_sari(vece, avr_full_offset(a->vrt), avr_full_offset(a->vrt),
```

Comment:
Instead of setting 1 bit of each element to 1/0 and then using ``sari`` to fill the other bits, how about:

```
TCGv_i64 ones = tcg_constant_i64((1 << elem_length) - 1);
```

And then replacing
```
tcg_gen_setcond_i64(TCG_COND_NE, t1, t1, tcg_constant_i64(0));
```
inside the ``for`` loop with
```
tcg_gen_movcond_i64(TCG_COND_NE, t1, t1, zero, ones, zero);
```

Reply:
How about:
```
    t0 = tcg_const_i64(0);
    t1 = tcg_temp_new_i64();
    zero = tcg_constant_i64(0);
    ones = tcg_constant_i64(MAKE_64BIT_MASK(0, elem_length));

    for (i = 1 << highest_bit; i > 1 << (highest_bit / 2); i >>= 1) {
        if (elem_length < 64) {
            tcg_gen_shli_i64(t0, t0, elem_length);
        }
        tcg_gen_ext_tl_i64(t1, cpu_gpr[a->vrb]);
        tcg_gen_andi_i64(t1, t1, i);
        if (elem_length < 64) {
            tcg_gen_movcond_i64(TCG_COND_NE, t1, t1, zero, ones, zero);
            tcg_gen_or_i64(t0, t0, t1);
        } else {
            tcg_gen_movcond_i64(TCG_COND_NE, t0, t1, zero, ones, zero);
        }
    }

    set_avr64(a->vrt, t0, true);

    (...)
```

Reply:
The effect on readability seems to make it not worth keeping `MO_64` with the other cases. I created an alternative PR in #42 using `movcond` and with `trans_MTVSRDM` out of `do_mtvsrm`.